Add influxdb support

tnachen opened the pull request.

aasssddd committed:
add influxdb_support

aasssddd committed:
update README

tnachen commented on metrics/influxdb.go:
```diff
@@ -0,0 +1,52 @@
+	gomet.Register(Requests, c[Requests])
+	c[Successes] = gomet.NewCounter()
+	gomet.Register(Successes, c[Successes])
+	s := gomet.NewExpDecaySample(1028, 0.015)
```
Why 0.015 and 1028?

aasssddd replied:
i just use the example from https://github.com/rcrowley/go-metrics
here is a paper describe what's it means, but i haven't study my self
[http://dimacs.rutgers.edu/~graham/pubs/papers/fwddecay.pdf](http://dimacs.rutgers.edu/~graham/pubs/papers/fwddecay.pdf)

tnachen replied:
We have to know what this means, the difference in load test results is going to impact all of our analysis and data.

tnachen commented on metrics/influxdb.go:
```diff
@@ -0,0 +1,52 @@
+
+// New :
+func (influx Influx) New() Influx {
+	c := make(map[string]gomet.Counter)
```
Please use more descriptive variable names, not c, h, s.

tnachen commented on metrics/influxdb.go:
```diff
@@ -0,0 +1,52 @@
+
+// Monitor : implement Metrics
+func (influx Influx) Monitor(opts *ServerOpts) {
+	go influxdb.InfluxDB(
```
This should only be called once right?

aasssddd replied:
yes, this will be called once

tnachen replied:
Let's make sure it can only get fired once then

tnachen commented on config.json:
```diff
@@ -0,0 +1,79 @@
+        "password": "root",
+        "db": "slow-cooker"
+    },
+    "job": [{
```
How is this job being used? I don't see your changes about this. If it's not used yet please take it out from this PR and just keep influx db config.

aasssddd replied:
this file is use for wrapper, i'll delete it in next commit

aasssddd committed:
1. delete unused file config.json
2. rename variable into more meaningful name

tnachen commented on README.md:
```diff
@@ -34,6 +34,11 @@ or:
 | `-interval`           | 10s       | How often to report stats to stdout. |
 | `-method`             | GET       | Determines which HTTP method to use when making the request. |
 | `-metric-addr`        | `<none>`  | Address to use when serving the Prometheus `/metrics` endpoint. No metrics are served if unset. Format is `host:port` or `:port`. |
+| `-use-promethus`      | `true`    | Use Promethus as metrics server |
```
How about just a flag --metrics-server-backend, and takes in promethus or influxdb. 
We might add other backends in the future (e.g: cassandra), and we can then just use one flag to support different backends

aasssddd replied:
looks good, on it.

tnachen commented on metrics/influxdb.go:
```diff
@@ -0,0 +1,52 @@
+}
+
+// CounterInc : implement Metrics
+func (influx Influx) CounterInc(name string) {
```
This writes to influx on every counter inc?
Prometheus and Influx works very differently, and we don't want to write on every data change. I think we should try to batch the writes instead of sending a request each stat change

aasssddd replied:
this method is only add 1 to counter, write to influx is periodically, it will send to influxdb every 5 seconds. should i make the write period to flag?

tnachen replied:
What if the run finishes before 5 seconds meet? Does the data get lost?

aasssddd replied:
unfortunately, yes. i'll try to dig into this issue.

aasssddd committed:
1. add lock to influx monitor daemon to make sure will run just only one goroutine
2. change command flag to -metric-server-backend
3. add lock to counterInc and histogram observe
4. shorten write to influx frequency to 1s

tnachen commented on metrics/influxdb.go:
```diff
@@ -0,0 +1,94 @@
+
+// CounterInc : implement Metrics
+func (influx Influx) CounterInc(name string) {
+	influx.counterLock.Lock()
```
This doesn't fix the data loss problem right?

aasssddd replied:
yes, this wouldn't fix this problem, i need to modify [go-metrics-influxdb](https://github.com/vrischmann/go-metrics-influxdb) and main.go to tell daemon sending all metrics immediately. 
will do today

tnachen commented on metrics/prometheus.go:
```diff
@@ -0,0 +1,71 @@
+// CounterInc : implement metrics
+func (p Prometheus) CounterInc(name string) {
+	p.Counter[name].Inc()
+
```
Remove extra space

tnachen commented on metrics/influxdb.go:
```diff
@@ -0,0 +1,94 @@
+		return
+	}
+	*influx.running = true
+	influx.threadLock.Lock()
```
You need to lock before you check running flag.
Also you need to unlock the threadLock too.

kozyraki commented:
I cannot tell from the code, but which metrics are you pushing there? The raw latencies captured or some of the statistics calculated?

aasssddd commented:
Hi @kozyraki, slow cooker have request count, successes count and statistics of latency (count, max, min, p50, p75, p95, p99, p999,p9999, stddev, variance).
these metrics will write to influxdb every 2 seconds. so you will get an aggregate record every 2 second.
If 2 seconds is not a proper period, you can run command with flag `-interval <value>`
Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".

aasssddd committed:
[Add]
1. add github.com/vrischmann/go-metrics-influxdb and dependencies to vendor
2. add SendMetricsNow to go-metrics-influxdb package
[metrics/influxdb.go, metrics/metrics.go, metrics/prometheus.go, main.go] add SendMetricsNow to Metrics interface, and implements for influxdb

[Bug fix]
1. [metrics/prometheus.go] remove redundant line
2. [metrics/influxdb.go] lock running flag before changing it, and call unlock at end of function

aasssddd commented on main.go:
```diff
@@ -372,6 +366,11 @@ func main() {
 			}
+
+			if *metricsServerBackend == ServerBackendInfluxDB {
+				met.SendMetricsNow()
```
fix data loss problem right here

aasssddd commented on vendor/github.com/vrischmann/go-metrics-influxdb/influxdb.go:
```diff
@@ -0,0 +1,219 @@
+var endSending = make(chan bool, 1)
+
+// SendMetricsNow Sends metrics immediately
+func SendMetricsNow() {
```
I add a SendMetricsNow function here, letting metrics send manually, and wait until metrics sent.

tnachen commented on metrics/influxdb.go:
```diff
@@ -0,0 +1,94 @@
+}
+
+// New :
+func (influx Influx) New() Influx {
```
This is not a typical go pattern.
You should have a package function like
func NewInflux(input.....) Influx {
  return Influx{
       .......
   }
}

So other places can just call NewInflux() to create a new influx object, instead of Influx{}.New()

tnachen commented on metrics/influxdb.go:
```diff
@@ -0,0 +1,94 @@
+	Counter                                map[string]gomet.Counter
+	Histogram                              map[string]gomet.Histogram
+	threadLock, counterLock, histogramLock *sync.Mutex
+	running                                *bool
```
Don't need a pointer to bool

tnachen commented on metrics/influxdb.go:
```diff
@@ -0,0 +1,94 @@
+	chRelease := make(chan bool, 1)
+
+	go func() {
+		influxdb.InfluxDB(
```
Shouldn't we get the error if it exits?

tnachen commented on metrics/influxdb.go:
```diff
@@ -0,0 +1,94 @@
+			opts.Password)
+
+		defer func() {
+			if r := recover(); r != nil {
```
What does recover do?

aasssddd replied:
it will regain control after panic occur, it just work like exception handling. 
see [Defer, Panic and Recover](https://blog.golang.org/defer-panic-and-recover)

tnachen replied:
Does influx call panic then?

aasssddd replied:
no need, panic will called when runtime error

tnachen commented on metrics/influxdb.go:
```diff
@@ -0,0 +1,94 @@
+	go func(influx *Influx) {
+		for {
+			switch {
+			case <-chRelease:
```
This seems too complicated for what we want to do. I don't think we ever need to stop Influx and restart again, so all we need is to have a simple bool to make sure we've started influx and we never need to start it again.
So basically just lock, run go func and set boolean.

aasssddd replied:
these code is to make sure the running status is set to false, when influx daemon goes down.
you really don't need it?

tnachen replied:
I think when influx is down we shouldn't even continue running as we won't even be able to store the stats. We should log and error out.

tnachen commented on metrics/metrics.go:
```diff
@@ -0,0 +1,26 @@
+	Monitor(opts *ServerOpts)
+	CounterInc(name string)
+	HistogramObserve(name string, data float64)
+	SendMetricsNow()
```
How about just calling this Sync()?

tnachen commented:
Can we also use glide or godep to manage vendor? So that we don't have to check in all the vendor code in our repo?

aasssddd committed:
[fix]
1. remove influx daemon's chRelease related code
2. rename SendMetricsNow into Sync
3. remove pointer to bool, pointer to whole struct instead.
4. remove vendor/ except go-metrics-influxdb
5. update Readme
6. update .gitignore
7. rename  Sync method

tnachen commented on main.go:
```diff
@@ -372,6 +366,11 @@ func main() {
 				hdrreport.PrintLatencySummary(globalHist)
 			}
+
+			if *metricsServerBackend == ServerBackendInfluxDB {
```
Let's just call Sync() without checking since Promethus is a no-op

tnachen commented on main.go:
```diff
@@ -354,12 +335,25 @@ func main() {
+
+	switch strings.ToLower(*metricsServerBackend) {
+	case ServerBackendPrometheus:
+		met = metrics.NewPrometheus()
```
call we have a more descriptive name? met is not obvious what this is.
How about metricsBackend?

tnachen commented on metrics/influxdb.go:
```diff
@@ -0,0 +1,81 @@
+
+// NewInflux :
+func NewInflux() *Influx {
+	influx := Influx{}
```
Fix this to use NewInflux()

tnachen commented on metrics/influxdb.go:
```diff
@@ -0,0 +1,81 @@
+		fmt.Println("monitor has already running")
+		return
+	}
+	influx.threadLock.Lock()
```
This is not correct, you need to lock before you check running flag, otherwise you can have two threads both checked running to be false, and they both start a goroutine.

tnachen commented on metrics/influxdb_test.go:
```diff
@@ -0,0 +1,35 @@
+	fmt.Println("run monitor on first instance")
+	influx.Monitor(&opts)
+	// call twice will return fail if goroutine is still running
+	fmt.Println("run twice on the same instance")
```
We don't need these Println in our test

tnachen committed:
Seperate out load test method

aasssddd committed:
[Fix]
modified:   main.go
Rename met into metricsBackend

modified:   metrics/influxdb.go
fix wrong lock logic problem

modified:   metrics/influxdb_test.go
remove unwanted println

aasssddd commented:
I'm using godep,
Because i add code to [go-metrics-influx](https://github.com/vrischmann/go-metrics-influxdb) for Sync method, so this package cannot control by godep or glide.
that's why vendor directory is not empty.

tnachen commented:
We also need to write out the latency stats and throughput stats

aasssddd committed:
add throughput histogram

aasssddd committed:
1. change NewSlidingTimeWindowSample signature
2. add histrogram-window-size flag to main.go

tnachen commented on metrics/sliding_window_sample.go:
```diff
@@ -0,0 +1,133 @@
+// Values : implements interface gomet.Sample
+func (s *SlidingWindowSample) Values() []int64 {
+	s.mutex.Lock()
+	s.mutex.Unlock()
```
defer?

aasssddd replied:
yes! thanks!

aasssddd committed:
1. Merge branch 'calibration' into add_influxdb_support
2. Fix SlidingTimeWindowSample deadlock problem

tnachen commented on metrics/sliding_time_window_sample.go:
```diff
@@ -0,0 +1,191 @@
+		lastTick: time.Now().UnixNano() * collisionBuffer,
+	}
+	if dur, err := time.ParseDuration(duration); err == nil {
+		sample.window = dur.Nanoseconds() * collisionBuffer
```
We shouldn't just hide the error, if there is one, then return the error.

tnachen replied:
also you should open a pr back to go metrics with this change

tnachen commented on metrics/sliding_time_window_sample.go:
```diff
@@ -0,0 +1,191 @@
+	defer s.mutex.Unlock()
+	oldTick := s.lastTick
+	tick := time.Now().UnixNano()
+	if tick-oldTick > 0 {
```
space between -

tnachen commented on metrics/sliding_time_window_sample.go:
```diff
@@ -0,0 +1,191 @@
+
+// Max : implements interface gomet.Sample
+func (s *SlidingTimeWindowSample) Max() int64 {
+	s.mutex.Lock()
```
Don't need to lock here, gomet internally already has mutex

aasssddd replied:
I'm implement a gomet.Sample interface to create a new Sampler. so i've to write thread-safe code myself.
reference: [here](https://github.com/rcrowley/go-metrics/blob/master/sample.go)

tnachen commented on metrics/sliding_time_window_sample.go:
```diff
@@ -0,0 +1,191 @@
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	return gomet.SampleMean(s.mappedValue())
```
Don't need to lock here, gomet internally already has mutex

tnachen commented on metrics/sliding_time_window_sample.go:
```diff
@@ -0,0 +1,191 @@
+func (s *SlidingTimeWindowSample) Min() int64 {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return gomet.SampleMin(s.mappedValue())
```
Don't need to lock here, gomet internally already has mutex

tnachen commented on metrics/sliding_time_window_sample.go:
```diff
@@ -0,0 +1,191 @@
+func (s *SlidingTimeWindowSample) Percentile(ps float64) float64 {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return gomet.SamplePercentile(s.mappedValue(), ps)
```
Don't need to lock here, gomet internally already has mutex

tnachen commented on metrics/sliding_time_window_sample.go:
```diff
@@ -0,0 +1,191 @@
+func (s *SlidingTimeWindowSample) Percentiles(ps []float64) []float64 {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return gomet.SamplePercentiles(s.mappedValue(), ps)
```
Don't need to lock here, gomet internally already has mutex

tnachen commented on metrics/sliding_time_window_sample.go:
```diff
@@ -0,0 +1,191 @@
+func (s *SlidingTimeWindowSample) Snapshot() gomet.Sample {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return gomet.NewSampleSnapshot(s.count, s.mappedValue())
```
Don't need to lock here, gomet internally already has mutex

tnachen replied:
Also other places

tnachen commented on metrics/sliding_window_sample.go:
```diff
@@ -0,0 +1,133 @@
+)
+
+// SlidingWindowSample : http://metrics.dropwizard.io/3.1.0/manual/core/#sliding-window-reservoirs
+type SlidingWindowSample struct {
```
Why do we need this and SlidingWindowTimeSample?

aasssddd replied:
SlidingWindowSample is not required for now, it just a spare, it gives a limited bucket size, and drop data from first if bucket is full.
SlidingWindowTimeSample is like SlidingWindowSample, it drop data if data is outdated.
Currently, i'm using SlidingWindowTimeSample for statistics latencies and throughput

aasssddd committed:
1. Extract RunLoad from main.go
2. Add API Server at server.go
3. Add route to slowcooker entry
